pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,220 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+from fnmatch import fnmatch
+
+from pootle.core.url_helpers import split_pootle_path
+from pootle_fs.utils import PathFilter
+from pootle_store.models import Store
+
+from .models import VirtualFolder, VirtualFolderTreeItem
+
+
+def update_vfolder_tree(vf, store):
+    """For a given VirtualFolder and Store update the VirtualFolderTreeItem
+    """
+    # Create missing VirtualFolderTreeItem tree structure for affected Stores
+    # after adding or unobsoleting Units.
+    created = False
+    try:
+        vfolder_treeitem = (
+            VirtualFolderTreeItem.objects.select_related("directory").get(
+                directory=store.parent, vfolder=vf))
+    except VirtualFolderTreeItem.DoesNotExist:
+        parts = split_pootle_path(store.parent.pootle_path)
+        path_parts = ["", parts[0], parts[1], vf.name]
+        if parts[2]:
+            path_parts.append(parts[2].strip("/"))
+        path_parts.append(parts[3])
+        pootle_path = "/".join(path_parts)
```

Comment:
It is not the first time I see this code. If it is going to stay after all vfolder-related PRs I would say refactor out.